### ISSUE TITLE
SVT: LG-261: MicroProfile Metrics 2.3 Implementation

### ISSUE THREAD

**Issue by tkulik** [OpenLiberty/open-liberty#10914]
Test mpMetrics 2.3 using a simple open liberty app.

**Comment by atrieder** [comment]
test

**Comment by atrieder** [comment]
### Initial Update

I am using the Open Liberty Getting Started application, which comes with `mpMetrics-2.2` enabled. After adding `mpMetrics-2.3` and `jaxrsMonitor-1.0` to the application's 'server.xml' file, I deployed the application to the latest Open Liberty Docker image (with the requisite feature assets copied in). Without any additional modifications, the below output is printed to the command line:
```
Launching defaultServer (WebSphere Application Server 20.0.0.3/wlp-1.0.38.cl200320200225-2113) on Eclipse OpenJ9 VM, version 1.8.0_242-b08 (en_US)
[AUDIT   ] CWWKE0001I: The server defaultServer has been launched.
[AUDIT   ] CWWKG0093A: Processing configuration drop-ins resource: /opt/ol/wlp/usr/servers/defaultServer/configDropins/defaults/keystore.xml
[AUDIT   ] CWWKG0093A: Processing configuration drop-ins resource: /opt/ol/wlp/usr/servers/defaultServer/configDropins/defaults/open-default-port.xml
[AUDIT   ] CWWKZ0058I: Monitoring dropins for applications.
[AUDIT   ] CWWKS4104A: LTPA keys created in 0.477 seconds. LTPA key file: /opt/ol/wlp/output/defaultServer/resources/security/ltpa.keys
[WARNING ] CWWKZ0014W: The application guide-getting-started could not be started as it could not be found at location guide-getting-started.war.
[AUDIT   ] CWWKT0016I: Web application available (default_host): http://172.18.0.1:9080/metrics/
[AUDIT   ] CWWKT0016I: Web application available (default_host): http://172.18.0.1:9080/health/
[AUDIT   ] CWWKT0016I: Web application available (default_host): http://172.18.0.1:9080/ibm/api/
[AUDIT   ] CWWKT0016I: Web application available (default_host): http://172.18.0.1:9080/application/
[AUDIT   ] CWWKZ0001I: Application application started in 0.685 seconds.
[AUDIT   ] CWWKF0012I: The server installed the following features: [cdi-2.0, distributedMap-1.0, jaxrs-2.1, jaxrsClient-2.1, jaxrsMonitor-1.0, jndi-1.0, json-1.0, jsonp-1.1, monitor-1.0, mpConfig-1.3, mpHealth-2.1, mpMetrics-2.3, servlet-4.0, ssl-1.0].
[AUDIT   ] CWWKF0011I: The defaultServer server is ready to run a smarter planet. The defaultServer server started in 6.675 seconds.
[AUDIT   ] CWPKI0803A: SSL certificate created in 1.706 seconds. SSL key file: /opt/ol/wlp/output/defaultServer/resources/security/key.p12
```

Overall, the metrics endpoint of the application behaves in the same manner as observed when running the application with `mpMetrics-2.2`. The next step is to observe new metric types specific to the new feature revision (i.e., simple timer) referenced [here](https://projects.eclipse.org/projects/technology.microprofile/releases/metrics-2.3).

**Comment by atrieder** [comment]
### Update

I added an instance of the simple timer metric type to the existing application. The output from the `/metrics/application` endpoint produced the following output in a web browser:
```
# TYPE application_getProperties_total counter
# HELP application_getProperties_total Number of times the JVM system properties are requested
application_getProperties_total 1
# TYPE application_io_openliberty_sample_system_SystemResource_getPropertiesTime_rate_per_second gauge
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_rate_per_second 0.008601344018868827
# TYPE application_io_openliberty_sample_system_SystemResource_getPropertiesTime_one_min_rate_per_second gauge
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_one_min_rate_per_second 0.03197594921593881
# TYPE application_io_openliberty_sample_system_SystemResource_getPropertiesTime_five_min_rate_per_second gauge
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_five_min_rate_per_second 0.1386081240172883
# TYPE application_io_openliberty_sample_system_SystemResource_getPropertiesTime_fifteen_min_rate_per_second gauge
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_fifteen_min_rate_per_second 0.17699033814381568
# TYPE application_io_openliberty_sample_system_SystemResource_getPropertiesTime_mean_seconds gauge
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_mean_seconds 0.003350378
# TYPE application_io_openliberty_sample_system_SystemResource_getPropertiesTime_max_seconds gauge
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_max_seconds 0.003350378
# TYPE application_io_openliberty_sample_system_SystemResource_getPropertiesTime_min_seconds gauge
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_min_seconds 0.003350378
# TYPE application_io_openliberty_sample_system_SystemResource_getPropertiesTime_stddev_seconds gauge
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_stddev_seconds 0.0
# TYPE application_io_openliberty_sample_system_SystemResource_getPropertiesTime_seconds summary
# HELP application_io_openliberty_sample_system_SystemResource_getPropertiesTime_seconds Time needed to get the JVM system properties
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_seconds_count 1
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_seconds{quantile="0.5"} 0.003350378
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_seconds{quantile="0.75"} 0.003350378
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_seconds{quantile="0.95"} 0.003350378
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_seconds{quantile="0.98"} 0.003350378
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_seconds{quantile="0.99"} 0.003350378
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_seconds{quantile="0.999"} 0.003350378
# TYPE application_io_openliberty_sample_system_SystemResource_getPropertiesSimpleTimer_total counter
# HELP application_io_openliberty_sample_system_SystemResource_getPropertiesSimpleTimer_total Like getPropertiesTime, but simpler!
application_io_openliberty_sample_system_SystemResource_getPropertiesSimpleTimer_total 1
# TYPE application_io_openliberty_sample_system_SystemResource_getPropertiesSimpleTimer_elapsedTime_seconds gauge
application_io_openliberty_sample_system_SystemResource_getPropertiesSimpleTimer_elapsedTime_seconds 0.004211102
```

The metric named "getPropertiesSimpleTimer" in the above output is the added simple timer metric. Functionality appears to be as expected.

**Comment by atrieder** [comment]
Since the new version of mpMetrics appears to work as expected, I think that this issue can be closed.

**Comment by atrieder** [comment]
Here is an example output from the /metrics endpoint of the Open Liberty Getting started application (from here: https://github.com/OpenLiberty/guide-getting-started). The application was running in the official Open Liberty Docker image with a modified build that included the latest `mpMetrics-2.3` and `jaxrsMonitor-1.0`  features. The metrics in the below output that contain base_REST_request showed up after we accessed the `/system/properties` endpoint in the application:
```
# TYPE base_cpu_processCpuLoad_percent gauge
# HELP base_cpu_processCpuLoad_percent Displays the 'recent cpu usage' for the Java Virtual Machine process.
base_cpu_processCpuLoad_percent 0.002939730614096964
# TYPE base_memory_maxHeap_bytes gauge
# HELP base_memory_maxHeap_bytes Displays the maximum amount of heap memory in bytes that can be used for memory management. This attribute displays -1 if the maximum heap memory size is undefined. This amount of memory is not guaranteed to be available for memory management if it is greater than the amount of committed memory. The Java virtual machine may fail to allocate memory even if the amount of used memory does not exceed this maximum size.
base_memory_maxHeap_bytes 5.36870912E8
# TYPE base_gc_time_seconds gauge
# HELP base_gc_time_seconds Displays the approximate accumulated collection elapsed time in milliseconds. This attribute displays -1 if the collection elapsed time is undefined for this collector. The Java virtual machine implementation may use a high resolution timer to measure the elapsed time. This attribute may display the same value even if the collection count has been incremented if the collection elapsed time is very short.
base_gc_time_seconds{name="scavenge"} 0.394
# TYPE base_gc_time_seconds gauge
# HELP base_gc_time_seconds Displays the approximate accumulated collection elapsed time in milliseconds. This attribute displays -1 if the collection elapsed time is undefined for this collector. The Java virtual machine implementation may use a high resolution timer to measure the elapsed time. This attribute may display the same value even if the collection count has been incremented if the collection elapsed time is very short.
base_gc_time_seconds{name="global"} 0.161
# TYPE base_memory_committedHeap_bytes gauge
# HELP base_memory_committedHeap_bytes Displays the amount of memory in bytes that is committed for the Java virtual machine to use. This amount of memory is guaranteed for the Java virtual machine to use.
base_memory_committedHeap_bytes 6.258688E7
# TYPE base_cpu_systemLoadAverage gauge
# HELP base_cpu_systemLoadAverage Displays the system load average for the last minute. The system load average is the sum of the number of runnable entities queued to the available processors and the number of runnable entities running on the available processors averaged over a period of time. The way in which the load average is calculated is operating system specific but is typically a damped time-dependent average. If the load average is not available, a negative value is displayed. This attribute is designed to provide a hint about the system load and may be queried frequently. The load average may be unavailable on some platform where it is expensive to implement this method.
base_cpu_systemLoadAverage 0.0
# TYPE base_gc_total counter
# HELP base_gc_total Displays the total number of collections that have occurred. This attribute lists -1 if the collection count is undefined for this collector.
base_gc_total{name="global"} 11
# TYPE base_gc_total counter
# HELP base_gc_total Displays the total number of collections that have occurred. This attribute lists -1 if the collection count is undefined for this collector.
base_gc_total{name="scavenge"} 132
# TYPE base_thread_daemon_count gauge
# HELP base_thread_daemon_count Displays the current number of live daemon threads.
base_thread_daemon_count 43
# TYPE base_cpu_availableProcessors gauge
# HELP base_cpu_availableProcessors Displays the number of processors available to the Java virtual machine. This value may change during a particular invocation of the virtual machine.
base_cpu_availableProcessors 4
# TYPE base_REST_request_total counter
# HELP base_REST_request_total The number of invocations and total response time of this RESTful resource method since the start of the server.
base_REST_request_total{class="io.openliberty.sample.system.SystemResource",method="getProperties"} 1
# TYPE base_REST_request_elapsedTime_seconds gauge
base_REST_request_elapsedTime_seconds{class="io.openliberty.sample.system.SystemResource",method="getProperties"} 0.10057605800000001
# TYPE base_memory_usedHeap_bytes gauge
# HELP base_memory_usedHeap_bytes Displays the amount of used heap memory in bytes.
base_memory_usedHeap_bytes 3.7988792E7
# TYPE base_thread_max_count gauge
# HELP base_thread_max_count Displays the peak live thread count since the Java virtual machine started or peak was reset. This includes daemon and non-daemon threads.
base_thread_max_count 58
# TYPE base_thread_count gauge
# HELP base_thread_count Displays the current number of live threads including both daemon and non-daemon threads.
base_thread_count 46
# TYPE base_classloader_loadedClasses_count gauge
# HELP base_classloader_loadedClasses_count Displays the number of classes that are currently loaded in the Java virtual machine.
base_classloader_loadedClasses_count 10147
# TYPE base_classloader_loadedClasses_total counter
# HELP base_classloader_loadedClasses_total Displays the total number of classes that have been loaded since the Java virtual machine has started execution.
base_classloader_loadedClasses_total 10150
# TYPE base_jvm_uptime_seconds gauge
# HELP base_jvm_uptime_seconds Displays the start time of the Java virtual machine in milliseconds. This attribute displays the approximate time when the Java virtual machine started.
base_jvm_uptime_seconds 996.3820000000001
# TYPE base_classloader_unloadedClasses_total counter
# HELP base_classloader_unloadedClasses_total Displays the total number of classes unloaded since the Java virtual machine has started execution.
base_classloader_unloadedClasses_total 3
# TYPE base_cpu_processCpuTime_seconds gauge
# HELP base_cpu_processCpuTime_seconds Displays the CPU time used by the process on which the Java virtual machine is running.
base_cpu_processCpuTime_seconds 38.716503
# TYPE vendor_threadpool_size gauge
# HELP vendor_threadpool_size The size of the thread pool.
vendor_threadpool_size{pool="Default_Executor"} 8
# TYPE vendor_session_create_total gauge
# HELP vendor_session_create_total The number of sessions that have logged in since this metric was enabled.
vendor_session_create_total{appname="default_host_metrics"} 1
# TYPE vendor_servlet_responseTime_total_seconds gauge
# HELP vendor_servlet_responseTime_total_seconds The total response time of this servlet since the start of the server.
vendor_servlet_responseTime_total_seconds{servlet="com_ibm_ws_microprofile_metrics_public_PublicMetricsRESTProxyServlet"} 0.18483848700000002
# TYPE vendor_servlet_responseTime_total_seconds gauge
# HELP vendor_servlet_responseTime_total_seconds The total response time of this servlet since the start of the server.
vendor_servlet_responseTime_total_seconds{servlet="guide_getting_started_io_openliberty_sample_system_SystemApplication"} 0.271364101
# TYPE vendor_session_activeSessions gauge
# HELP vendor_session_activeSessions The number of concurrently active sessions. (A session is active if the product is currently processing a request that uses that user session).
vendor_session_activeSessions{appname="default_host_metrics"} 0
# TYPE vendor_session_invalidated_total counter
# HELP vendor_session_invalidated_total The number of sessions that have logged out since this metric was enabled.
vendor_session_invalidated_total{appname="default_host_metrics"} 0
# TYPE vendor_session_invalidatedbyTimeout_total counter
# HELP vendor_session_invalidatedbyTimeout_total The number of sessions that have logged out by timeout since this metric was enabled.
vendor_session_invalidatedbyTimeout_total{appname="default_host_metrics"} 0
# TYPE vendor_servlet_request_total counter
# HELP vendor_servlet_request_total The number of visits to this servlet since the start of the server.
vendor_servlet_request_total{servlet="com_ibm_ws_microprofile_metrics_public_PublicMetricsRESTProxyServlet"} 20
# TYPE vendor_session_liveSessions gauge
# HELP vendor_session_liveSessions The number of users that are currently logged in.
vendor_session_liveSessions{appname="default_host_metrics"} 1
# TYPE vendor_servlet_request_total counter
# HELP vendor_servlet_request_total The number of visits to this servlet since the start of the server.
vendor_servlet_request_total{servlet="guide_getting_started_io_openliberty_sample_system_SystemApplication"} 1
# TYPE vendor_threadpool_activeThreads gauge
# HELP vendor_threadpool_activeThreads The number of threads that are running tasks.
vendor_threadpool_activeThreads{pool="Default_Executor"} 1
# TYPE application_getProperties_total counter
# HELP application_getProperties_total Number of times the JVM system properties are requested
application_getProperties_total 1
# TYPE application_io_openliberty_sample_system_SystemResource_getPropertiesTime_rate_per_second gauge
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_rate_per_second 0.02932089499558698
# TYPE application_io_openliberty_sample_system_SystemResource_getPropertiesTime_one_min_rate_per_second gauge
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_one_min_rate_per_second 0.1318481260400888
# TYPE application_io_openliberty_sample_system_SystemResource_getPropertiesTime_five_min_rate_per_second gauge
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_five_min_rate_per_second 0.18400888292586465
# TYPE application_io_openliberty_sample_system_SystemResource_getPropertiesTime_fifteen_min_rate_per_second gauge
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_fifteen_min_rate_per_second 0.1945208954232697
# TYPE application_io_openliberty_sample_system_SystemResource_getPropertiesTime_mean_seconds gauge
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_mean_seconds 0.0035151020000000004
# TYPE application_io_openliberty_sample_system_SystemResource_getPropertiesTime_max_seconds gauge
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_max_seconds 0.0035151020000000004
# TYPE application_io_openliberty_sample_system_SystemResource_getPropertiesTime_min_seconds gauge
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_min_seconds 0.0035151020000000004
# TYPE application_io_openliberty_sample_system_SystemResource_getPropertiesTime_stddev_seconds gauge
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_stddev_seconds 0.0
# TYPE application_io_openliberty_sample_system_SystemResource_getPropertiesTime_seconds summary
# HELP application_io_openliberty_sample_system_SystemResource_getPropertiesTime_seconds Time needed to get the JVM system properties
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_seconds_count 1
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_seconds{quantile="0.5"} 0.0035151020000000004
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_seconds{quantile="0.75"} 0.0035151020000000004
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_seconds{quantile="0.95"} 0.0035151020000000004
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_seconds{quantile="0.98"} 0.0035151020000000004
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_seconds{quantile="0.99"} 0.0035151020000000004
application_io_openliberty_sample_system_SystemResource_getPropertiesTime_seconds{quantile="0.999"} 0.0035151020000000004
```
